File: backend/app/schemas/ai.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, validator

from .events import EventRead


class ParticipantProfile(BaseModel):
    name: str
    bio: Optional[str] = None
    interests: List[str] = Field(default_factory=list)
    shared_activities: List[str] = Field(default_factory=list)

    @validator("interests", "shared_activities", pre=True, always=True)
    def _coerce_list(cls, value):
        if value is None:
            return []
        if isinstance(value, str):
            return [item.strip() for item in value.split(",") if item.strip()]
        return value


class MatchInsightRequest(BaseModel):
    participants: List[ParticipantProfile] = Field(..., min_items=2)
    shared_interests: List[str] = Field(default_factory=list)
    shared_activities: List[str] = Field(default_factory=list)
    mood: Optional[str] = None
    location: Optional[str] = None


class MatchInsightResponse(BaseModel):
    match_id: str
    summary_text: str
    generated_at: datetime
    cached: bool = False
    moderation_applied: bool = False

    class Config:
        orm_mode = True


class DateIdeaWindow(BaseModel):
    start: datetime
    end: datetime

    @validator("end")
    def end_after_start(cls, value, values):
        start = values.get("start")
        if start and value <= start:
            raise ValueError("availability window end must be after start")
        return value


class DateIdeaRequest(BaseModel):
    match_id: str = Field(..., description="Identifier for the match or group connection")
    shared_interests: List[str] = Field(default_factory=list)
    location: Optional[str] = Field(default=None, description="Campus or city anchor point")
    availability_window: Optional[DateIdeaWindow] = None
    mood: Optional[str] = None
    weather: Optional[str] = None
    participants: List[ParticipantProfile] = Field(default_factory=list)


class DateIdea(BaseModel):
    id: int
    match_id: str
    title: str
    description: str
    location: Optional[str]
    idea_rank: int
    generated_at: datetime
    expires_at: datetime

    class Config:
        orm_mode = True


class DateIdeasResponse(BaseModel):
    match_id: str
    ideas: List[DateIdea]
    cached: bool = False
    generated_at: datetime


class EventFilterDateRange(BaseModel):
    start: Optional[datetime] = None
    end: Optional[datetime] = None


class EventFilters(BaseModel):
    date_range: Optional[EventFilterDateRange] = None
    location: Optional[str] = None
    category: Optional[str] = None
    keywords: List[str] = Field(default_factory=list)


class EventNLPResponse(BaseModel):
    query: str
    filters: EventFilters
    events: List[EventRead]
    cached: bool = False
    interpreted_query: str
    generated_at: datetime

File: frontend/src/services/ai.ts
import api from '../lib/api';

export interface ParticipantProfile {
  name: string;
  bio?: string | null;
  interests?: string[];
  shared_activities?: string[];
}

export interface MatchInsightRequest {
  participants: ParticipantProfile[];
  shared_interests?: string[];
  shared_activities?: string[];
  mood?: string;
  location?: string;
}

export interface MatchInsightResponse {
  match_id: string;
  summary_text: string;
  generated_at: string;
  cached: boolean;
  moderation_applied: boolean;
}

export interface DateIdea {
  id: number;
  match_id: string;
  title: string;
  description: string;
  location?: string | null;
  idea_rank: number;
  generated_at: string;
  expires_at: string;
}

export interface DateIdeaRequest {
  match_id: string;
  shared_interests?: string[];
  location?: string;
  availability_window?: {
    start: string;
    end: string;
  } | null;
  mood?: string;
  weather?: string;
  participants?: ParticipantProfile[];
}

export interface DateIdeasResponse {
  match_id: string;
  ideas: DateIdea[];
  cached: boolean;
  generated_at: string;
}

export interface EventFilterDateRange {
  start?: string | null;
  end?: string | null;
}

export interface EventFilters {
  date_range?: EventFilterDateRange | null;
  location?: string | null;
  category?: string | null;
  keywords?: string[];
}

export interface EventRecord {
  id: string | number;
  title: string;
  description?: string | null;
  location: string;
  category: string;
  start_time: string;
  end_time: string;
  tags?: string[];
  imageUrl?: string | null;
  lat?: number | null;
  lng?: number | null;
  interest_count?: number;
  viewer_interest?: boolean;
}

export interface EventNlpResponse {
  query: string;
  filters: EventFilters;
  events: EventRecord[];
  cached: boolean;
  interpreted_query: string;
  generated_at: string;
}

export const aiApi = {
  getMatchInsight: (matchId: string, refresh = false) =>
    api.get<MatchInsightResponse>(`/matches/${matchId}/insight`, {
      params: refresh ? { refresh: true } : undefined,
    }),

  generateMatchInsight: (matchId: string, payload: MatchInsightRequest) =>
    api.post<MatchInsightResponse>(`/matches/${matchId}/insight`, payload),

  generateDateIdeas: (payload: DateIdeaRequest) =>
    api.post<DateIdeasResponse>('/ideas', payload),

  fetchDateIdeas: (matchId: string, refresh = false) =>
    api.get<DateIdeasResponse>('/ideas', {
      params: {
        match_id: matchId,
        ...(refresh ? { refresh: true } : {}),
      },
    }),

  searchEvents: (query: string, viewerId?: string, refresh = false) =>
    api.get<EventNlpResponse>('/events/nlp-search', {
      params: {
        q: query,
        ...(refresh ? { refresh: true } : {}),
        ...(viewerId ? { viewer_id: viewerId } : {}),
      },
    }),
};
